osu.Game/Online/Chat/ChannelManager.cs

```diff
@@ -440,24 +444,44 @@ public void LeaveChannel(Channel channel)
         /// </summary>
         public void JoinLastClosedChannel()
         {
-            if (closedChannels.Count <= 0)
+            if (closedChannelIds.Count <= 0)
             {
                 return;
             }
 
-            Channel lastClosedChannel = closedChannels.Last();
+            // This nested loop could be eliminated if a check was added so that
+            // when the code opens a channel it removes from the closedChannel list.
+            // However, this would require adding an O(|closeChannels|) work operation
+            // every time the user joins a channel, which would make joining a channel
+            // slower. I wanted to centralize all major slowdowns so they
+            // can only occur if the user actually decides to use this feature.
+            while (closedChannelIds.Count != 0)
+            {
+                long lastClosedChannelId = closedChannelIds.Last();
+                closedChannelIds.RemoveAt(closedChannelIds.Count - 1);
 
-            closedChannels.Remove(lastClosedChannel);
+                bool lookupCondition(Channel ch) =>
+                    ch.Type == ChannelType.PM ? ch.Users.Any(u => u.Id == lastClosedChannelId)
+                                              : ch.Id == lastClosedChannelId;
 
-            // If the user already joined the channel, try the next
-            // channel in the list
-            if (joinedChannels.IndexOf(lastClosedChannel) >= 0)
-            {
-                JoinLastClosedChannel();
-            }
-            else
-            {
-                JoinChannel(lastClosedChannel);
+                // If the user hasn't already joined the channel, try to join it
+                if (joinedChannels.FirstOrDefault(lookupCondition) == null)
+                {
+                    Channel lastClosedChannel = AvailableChannels.FirstOrDefault(lookupCondition);
+
+                    if (lastClosedChannel != null)
+                    {
+                        CurrentChannel.Value = JoinChannel(lastClosedChannel);
+                    }
+                    else
+                    {
+                        // Try to get User to open PM chat
+                        var req = new GetUserRequest(lastClosedChannelId);
```

Comment:
This does make an API call which is kind of iffy if this shortcut is called multiple times. I see that there's a UserLookupCache - we could use that to reduce the number of API calls?